src/java.base/share/classes/sun/util/locale/provider/BreakIteratorProviderImpl.java

```diff
@@ -326,7 +326,7 @@ static final class CharacterIteratorCharSequence implements CharSequence {
 
         @Override
         public int length() {
-            return src.getEndIndex() - src.getBeginIndex();
+            return src.getEndIndex();
```

Comment:
Could the issue be somewhere else?  I mean it feels correct for length = endIndex - beginIndex.

Reply:
It's somewhat confusing as this class adapts `CharacterIterator` into `CharSequence` which are similar to each other (thus the bug). Those `beginIndex`/`endIndex` designate the range in the source `CharacterIterator`, and this `length()` method should return the entire text length of the `CharSequence` nevertheless, thus it should start from `0` to `endIndex`.

Reply:
Yeah, I saw there's a mismatch between the src and limit in this call that led to the index check Exception:
286             for (int b = ci.getBeginIndex(); b < end;) {
287                 boundaries.add(b);
288                 b = Grapheme.nextBoundary(text, b, end);
289             }

and nextBoundary could walk through the entire CharSequence. Maybe it's worth a note to the length() method.

Reply:
Thanks, Joe. Added comment to the method.

Reply:
OK yeah this is really confusing. One might ask a similar question in `charAt` about why it doesn't call

    src.setIndex(index + src.getBeginIndex());

The answer is that this is a special-purpose `CharSequence` that represents characters in the index range [0..endIndex) of the underlying `CharacterIterator`, **even if** that `CharacterIterator` represents the subrange of some string. As Joe noted, the calling code in `GraphemeBreakIterator` takes care to ensure that only valid indexes into the `src` are used.

I think the comment on `length` helps a little bit but maybe a class-level comment would be better, since it applies to the whole model of this special `CharSequence` and not just the `length` method. Just a sentence or two is sufficient. Feel free to crib from the above.

Reply:
Thanks, Stuart. I used the paragraph almost intact for the class description 😀

Reply:
OK looks good.